### PR TITLE
Refactor Semaphore for readability

### PR DESCRIPTION
**Pull request by bplommer**

### CODE REVIEW

File: concurrent/shared/src/main/scala/cats/effect/concurrent/Semaphore.scala [hunk 1]
```diff
@@ -129,9 +129,13 @@ object Semaphore {
...
     else F.unit
 
+  private final case class Request[F[_]](n: Long, gate: Deferred[F, Unit])
+
   // A semaphore is either empty, and there are number of outstanding acquires (Left)
   // or it is non-empty, and there are n permits available (Right)
-  private type State[F[_]] = Either[Queue[(Long, Deferred[F, Unit])], Long]
+  private type State[F[_]] = Either[Queue[Request[F]], Long]
+
+  private final case class Promise[F[_]](await: F[Unit], release: F[Unit])
```
Comment (RaasAhsan) [inline concurrent/shared/src/main/scala/cats/effect/concurrent/Semaphore.scala]: This doesn't really feel like a `Promise` to me (there were a ton of discussions about naming something `promise` which are still fresh in my mind :D ), maybe something more like `Acquire` or `Permit`?

**Review by RaasAhsan — APPROVED** [review]
One comment about naming, but this looks awesome! Named data types always makes these state machines way easier to follow

File: concurrent/shared/src/main/scala/cats/effect/concurrent/Semaphore.scala (hunk 1 above)
Comment (bplommer) [reply concurrent/shared/src/main/scala/cats/effect/concurrent/Semaphore.scala]: `Permit` doesn't seem right - when you get this you don't yet have a permit, just a way to manage one. Permits themselves don't have a first-class representation. `Acquire` feels vague. Maybe `PermitToken` or `Registration`?

Comment (SystemFw) [reply concurrent/shared/src/main/scala/cats/effect/concurrent/Semaphore.scala]: `Permit` seems ok to me, since it will read like `permit.await`

Comment (bplommer) [reply concurrent/shared/src/main/scala/cats/effect/concurrent/Semaphore.scala]: I'll change it to `Permit` since that seems to be the consensus 🙂

**Pull request merged**